File: LittleBigMouse.Plugins/LittleBigMouse.Plugins.Core/Persistence/LayoutDtoMapper.cs
#nullable enable
using System.Linq;
using LittleBigMouse.DisplayLayout.Dimensions;
using LittleBigMouse.DisplayLayout.Monitors;

namespace LittleBigMouse.Plugins.Persistence;

/// <summary>
/// The whole model↔DTO mapping, once, for every platform. <c>Apply</c> reads a DTO into
/// the live model (null field = absent from the store = keep the live value),
/// <c>ToDto</c> writes the model back. The two directions stay in this one file on
/// purpose: adding a persisted field means one DTO property (LayoutDtos.cs) and its two
/// mapping lines here — no per-OS code, so Windows/Linux can no longer drift apart.
/// <para>
/// Nothing here interprets history: reading a value that means something different
/// depending on when it was written goes through <see cref="LayoutMigrations"/>.
/// </para>
/// </summary>
public static class LayoutDtoMapper
{
    //==================//
    // DTO -> model     //
    //==================//

    public static void Apply(ILayoutOptions o, GlobalOptionsDto? dto)
    {
        if (dto == null) return;

        o.DaemonPort = dto.DaemonPort ?? o.DaemonPort;
        o.Priority = dto.Priority ?? o.Priority;
        o.PriorityUnhooked = dto.PriorityUnhooked ?? o.PriorityUnhooked;
        o.HomeCinema = dto.HomeCinema ?? o.HomeCinema;
        o.Pinned = dto.Pinned ?? o.Pinned;
        o.AutoUpdate = dto.AutoUpdate ?? o.AutoUpdate;
        o.StartMinimized = dto.StartMinimized ?? o.StartMinimized;
        o.StartElevated = dto.StartElevated ?? o.StartElevated;
        o.DebugTools = dto.DebugTools ?? o.DebugTools;
        o.VcpControl = dto.VcpControl ?? o.VcpControl;
        o.ShowMonitorActionWarning = dto.ShowMonitorActionWarning ?? o.ShowMonitorActionWarning;
        o.BorderValues = dto.BorderValues ?? o.BorderValues;
        o.RescueShortcut = dto.RescueShortcut ?? o.RescueShortcut;
        o.HideTrayIcon = dto.HideTrayIcon ?? o.HideTrayIcon;
    }

    public static void Apply(ILayoutOptions o, LayoutOptionsDto? dto)
    {
        if (dto == null) return;

        o.AllowOverlaps = dto.AllowOverlaps ?? o.AllowOverlaps;
        o.AllowDiscontinuity = dto.AllowDiscontinuity ?? o.AllowDiscontinuity;
        o.Algorithm = dto.Algorithm ?? o.Algorithm;
        o.MaxTravelDistance = dto.MaxTravelDistance ?? o.MaxTravelDistance;
        o.FreelookCheckInterval = dto.FreelookCheckInterval ?? o.FreelookCheckInterval;
        o.FreelookEnabled = dto.FreelookEnabled ?? o.FreelookEnabled;
        o.LoopX = dto.LoopX ?? o.LoopX;
        o.LoopY = dto.LoopY ?? o.LoopY;
        o.Enabled = dto.Enabled ?? o.Enabled;
        o.AdjustPointer = dto.AdjustPointer ?? o.AdjustPointer;
        o.AdjustSpeed = dto.AdjustSpeed ?? o.AdjustSpeed;

        // READ-ONLY legacy: the app-level values are applied first, and a layout that
        // still carries its own overrides them so nobody's priority changes on upgrade.
        // Nothing writes them back (see ToDto below), so this ends at the first save.
        o.Priority = dto.Priority ?? o.Priority;
        o.PriorityUnhooked = dto.PriorityUnhooked ?? o.PriorityUnhooked;
    }

    public static void Apply(PhysicalMonitorModel model, ModelDto dto)
    {
        var fixedRatio = model.PhysicalSize.FixedAspectRatio;
        model.PhysicalSize.FixedAspectRatio = false;

        model.PhysicalSize.TopBorder = dto.Borders?.Top ?? model.PhysicalSize.TopBorder;
        model.PhysicalSize.RightBorder = dto.Borders?.Right ?? model.PhysicalSize.RightBorder;
        model.PhysicalSize.BottomBorder = dto.Borders?.Bottom ?? model.PhysicalSize.BottomBorder;
        model.PhysicalSize.LeftBorder = dto.Borders?.Left ?? model.PhysicalSize.LeftBorder;

        // Placeholder sizes (#419) and pre-5.4.1 oriented sizes (#507) — see LayoutMigrations.
        var (width, height) = LayoutMigrations.StoredModelSize(
            model.PhysicalSize.Width, model.PhysicalSize.Height, dto.Width, dto.Height);

        if (width is { } w) model.PhysicalSize.Width = w;
        if (height is { } h) model.PhysicalSize.Height = h;

        model.PhysicalSize.FixedAspectRatio = fixedRatio;

        if (!string.IsNullOrEmpty(dto.PnpName)) model.PnpDeviceName = dto.PnpName;
    }

    public static void Apply(PhysicalMonitor monitor, MonitorDto dto)
    {
        foreach (var source in monitor.Sources.Items)
        {
            if (dto.Sources == null || !dto.Sources.TryGetValue(source.Source.Id, out var s)) continue;

            // Detached sources restore their stored pixel geometry (nothing current to
            // keep); attached ones keep the live geometry, the store is just a backup.
            if (!source.Source.AttachedToDesktop)
            {
                source.Source.InPixel.Set(new HLab.Geo.Rect(
                    new HLab.Geo.Point(s.PixelX ?? source.Source.InPixel.X, s.PixelY ?? source.Source.InPixel.Y),
                    new HLab.Geo.Size(s.PixelWidth ?? source.Source.InPixel.Width, s.PixelHeight ?? source.Source.InPixel.Height)));
                source.Source.Orientation = s.Orientation ?? source.Source.Orientation;
            }

            if (dto.ActiveSource != null && source.Source.Id == dto.ActiveSource)
                monitor.ActiveSource = source;
        }

        if (dto.XLocationInMm is { } x) { monitor.DepthProjection.X = x; monitor.Placed = true; }
        if (dto.YLocationInMm is { } y) { monitor.DepthProjection.Y = y; monitor.Placed = true; }

        monitor.DepthRatio.X = dto.PhysicalRatioX ?? monitor.DepthRatio.X;
        monitor.DepthRatio.Y = dto.PhysicalRatioY ?? monitor.DepthRatio.Y;

        // The edge length is needed to convert a stored whole-edge resistance into
        // the section that now expresses it.
        var acrossMm = monitor.DepthProjection.Width;
        var downMm = monitor.DepthProjection.Height;

        Apply(monitor.BorderResistance.Left, dto.BorderResistance?.Left, downMm);
        Apply(monitor.BorderResistance.Top, dto.BorderResistance?.Top, acrossMm);
        Apply(monitor.BorderResistance.Right, dto.BorderResistance?.Right, downMm);
        Apply(monitor.BorderResistance.Bottom, dto.BorderResistance?.Bottom, acrossMm);

        monitor.ExcludedFromLayout = dto.ExcludedFromLayout ?? monitor.ExcludedFromLayout;

        // Per-monitor bezel borders load whatever the current mode is, so switching to
        // PerMonitor is live (no restart required). Stored values only exist once the
        // user edited them in PerMonitor mode: until then Borders keep mirroring the
        // live model values, so the FIRST switch starts from the monitor's current
        // PerModel borders.
        if (dto.Borders != null)
        {
            monitor.Borders.Left = dto.Borders.Left ?? monitor.Borders.Left;
            monitor.Borders.Top = dto.Borders.Top ?? monitor.Borders.Top;
            monitor.Borders.Right = dto.Borders.Right ?? monitor.Borders.Right;
            monitor.Borders.Bottom = dto.Borders.Bottom ?? monitor.Borders.Bottom;
            monitor.BordersCustomized = true;
        }
    }

    public static void Apply(BorderSide side, BorderSideDto? dto, double edgeLengthMm)
    {
        if (dto == null) return;

        side.Sections.Edit(list =>
        {
            list.Clear();

            foreach (var s in LayoutMigrations.Sections(dto, edgeLengthMm))
            {
                list.Add(new BorderSection
                {
                    From = s.From ?? 0,
                    To = s.To ?? 0,
                    Move = s.Move ?? 0,
                    MoveBlock = s.MoveBlock ?? false,
                    Drag = s.Drag ?? 0,
                    DragBlock = s.DragBlock ?? false
                });
            }
        });
    }

    //==================//
    // model -> DTO     //
    //==================//

    /// <summary>
    /// App-level options. <paramref name="excludedDefaultsVersion"/> is not part of the
    /// options model: it is read at load time and round-tripped through every write so the
    /// one-time exclusion-list top-up stays one-time (see <see cref="ExcludedListPersistence"/>).
    /// </summary>
    public static GlobalOptionsDto ToGlobalOptionsDto(ILayoutOptions o, int? excludedDefaultsVersion) => new()
    {
        DaemonPort = o.DaemonPort,
        Priority = o.Priority,
        PriorityUnhooked = o.PriorityUnhooked,
        HomeCinema = o.HomeCinema,
        Pinned = o.Pinned,
        AutoUpdate = o.AutoUpdate,
        StartMinimized = o.StartMinimized,
        StartElevated = o.StartElevated,
        DebugTools = o.DebugTools,
        VcpControl = o.VcpControl,
        ShowMonitorActionWarning = o.ShowMonitorActionWarning,
        BorderValues = o.BorderValues,
        RescueShortcut = o.RescueShortcut,
        HideTrayIcon = o.HideTrayIcon,
        ExcludedDefaultsVersion = excludedDefaultsVersion
    };

    public static LayoutDto ToLayoutDto(MonitorsLayout layout) => new()
    {
        Options = ToDto(layout.Options),
        Monitors = layout.PhysicalMonitors.ToDictionary(m => m.Id, ToDto)
    };

    // Priority/PriorityUnhooked are deliberately absent: they are app-level options that
    // once lived here too, and both locations load into the SAME options property. Writing
    // the property back to both made a layout's leftover value the app-level one at the
    // next save, and gave a copy of it to every other layout after that. They are written
    // to the app level only; the layout copy is read as a fallback and then dropped.
    public static LayoutOptionsDto ToDto(ILayoutOptions o) => new()
    {
        AllowOverlaps = o.AllowOverlaps,
        AllowDiscontinuity = o.AllowDiscontinuity,
        Algorithm = o.Algorithm,
        MaxTravelDistance = o.MaxTravelDistance,
        FreelookCheckInterval = o.FreelookCheckInterval,
        FreelookEnabled = o.FreelookEnabled,
        LoopX = o.LoopX,
        LoopY = o.LoopY,
        Enabled = o.Enabled,
        AdjustPointer = o.AdjustPointer,
        AdjustSpeed = o.AdjustSpeed
    };

    // Sections only: the per-edge resistance is gone, and writing it back would
    // resurrect it on the next load through the migration path. Always emitted, even
    // empty, so deleting the last section clears what the store still holds.
    public static BorderSideDto ToDto(BorderSide side) => new()
    {
        Sections = side.Sections.Items.Count == 0
            ? null
            : [.. side.Sections.Items.Select(s => new BorderSectionDto
            {
                From = s.From,
                To = s.To,
                Move = s.Move,
                MoveBlock = s.MoveBlock,
                Drag = s.Drag,
                DragBlock = s.DragBlock
            })]
    };

    public static MonitorDto ToDto(PhysicalMonitor monitor) => new()
    {
        XLocationInMm = monitor.DepthProjection.X,
        YLocationInMm = monitor.DepthProjection.Y,
        PhysicalRatioX = monitor.DepthRatio.X,
        PhysicalRatioY = monitor.DepthRatio.Y,
        BorderResistance = new BorderResistanceDto
        {
            Left = ToDto(monitor.BorderResistance.Left),
            Top = ToDto(monitor.BorderResistance.Top),
            Right = ToDto(monitor.BorderResistance.Right),
            Bottom = ToDto(monitor.BorderResistance.Bottom)
        },
        // Stored whatever the current mode is (they must survive a Save() made in
        // PerModel mode), but only once the monitor owns them: uncustomized monitors
        // keep mirroring the model and store nothing.
        Borders = monitor.BordersCustomized
            ? new BordersDto
            {
                Left = monitor.Borders.Left,
                Top = monitor.Borders.Top,
                Right = monitor.Borders.Right,
                Bottom = monitor.Borders.Bottom
            }
            : null,
        ActiveSource = monitor.ActiveSource.Source.Id,
        SerialNumber = monitor.SerialNumber,
        ExcludedFromLayout = monitor.ExcludedFromLayout,
        Sources = monitor.Sources.Items
            .Where(s => s.Source.AttachedToDesktop)
            .ToDictionary(s => s.Source.Id, s => new SourceDto
            {
                PixelX = s.Source.InPixel.X,
                PixelY = s.Source.InPixel.Y,
                PixelWidth = s.Source.InPixel.Width,
                PixelHeight = s.Source.InPixel.Height,
                Orientation = s.Source.Orientation,
                DisplayName = s.Source.DisplayName,
                Primary = s.Source.Primary
            })
    };

    public static ModelDto ToDto(PhysicalMonitorModel model) => new()
    {
        Width = model.PhysicalSize.Width,
        Height = model.PhysicalSize.Height,
        Borders = new BordersDto
        {
            Left = model.PhysicalSize.LeftBorder,
            Top = model.PhysicalSize.TopBorder,
            Right = model.PhysicalSize.RightBorder,
            Bottom = model.PhysicalSize.BottomBorder
        },
        PnpName = model.PnpDeviceName
    };
}

File: rust/crates/lbm-store/src/layout_dto_mapper.rs
//! The whole model↔DTO mapping — port of `Persistence/LayoutDtoMapper.cs`.
//!
//! The `apply_*` functions read a DTO into the live model (a `None` field is absent
//! from the store: the live value is kept), the `to_*` functions write the model back.
//! Both directions stay in this one file on purpose: adding a persisted field means one
//! DTO member ([`layout_dtos`](crate::layout_dtos)) and its two mapping lines here.
//!
//! Nothing here interprets history: reading a value that means something different
//! depending on when it was written goes through [`layout_migrations`].
//!
//! The model is changed through the layout's edit functions, one per C# property
//! write, so every write has the side effects it has in C# (saved flags, border
//! mirroring, republished values). The engine marks everything saved after a load.

use indexmap::IndexMap;
use lbm_layout::geo::{Point, Rect, Size, Thickness};
use lbm_layout::model::{
    BorderSection, BorderSide, Layout, LayoutOptions, Monitor, MonitorModel, Ratio,
};

use crate::layout_dtos::{
    BorderResistanceDto, BorderSectionDto, BorderSideDto, BordersDto, GlobalOptionsDto, LayoutDto,
    LayoutOptionsDto, ModelDto, MonitorDto, SourceDto,
};
use crate::layout_migrations;

//==================//
// DTO -> model     //
//==================//

/// `o.X = dto.X ?? o.X`.
fn keep<T: Clone>(field: &mut T, stored: &Option<T>) {
    if let Some(value) = stored {
        *field = value.clone();
    }
}

/// C# `Apply(ILayoutOptions, GlobalOptionsDto?)`: the app-level options.
pub fn apply_global_options(o: &mut LayoutOptions, dto: Option<&GlobalOptionsDto>) {
    let Some(dto) = dto else { return };

    keep(&mut o.priority, &dto.priority);
    keep(&mut o.priority_unhooked, &dto.priority_unhooked);
    keep(&mut o.home_cinema, &dto.home_cinema);
    keep(&mut o.pinned, &dto.pinned);
    keep(&mut o.auto_update, &dto.auto_update);
    keep(&mut o.start_minimized, &dto.start_minimized);
    keep(&mut o.start_elevated, &dto.start_elevated);
    keep(&mut o.debug_tools, &dto.debug_tools);
    keep(&mut o.experimental_features, &dto.experimental_features);
    keep(&mut o.vcp_control, &dto.vcp_control);
    keep(
        &mut o.show_monitor_action_warning,
        &dto.show_monitor_action_warning,
    );
    keep(&mut o.border_values, &dto.border_values);
    keep(&mut o.rescue_shortcut, &dto.rescue_shortcut);
    keep(&mut o.hide_tray_icon, &dto.hide_tray_icon);
}

/// C# `Apply(ILayoutOptions, LayoutOptionsDto?)`: the per-layout options.
pub fn apply_layout_options(o: &mut LayoutOptions, dto: Option<&LayoutOptionsDto>) {
    let Some(dto) = dto else { return };

    keep(&mut o.allow_overlaps, &dto.allow_overlaps);
    keep(&mut o.allow_discontinuity, &dto.allow_discontinuity);
    keep(&mut o.algorithm, &dto.algorithm);
    keep(&mut o.minimal_edge_overlap, &dto.minimal_edge_overlap);
    keep(&mut o.max_travel_distance, &dto.max_travel_distance);
    keep(&mut o.freelook_check_interval, &dto.freelook_check_interval);
    keep(&mut o.freelook_enabled, &dto.freelook_enabled);
    keep(&mut o.loop_x, &dto.loop_x);
    keep(&mut o.loop_y, &dto.loop_y);
    keep(&mut o.enabled, &dto.enabled);
    keep(&mut o.adjust_pointer, &dto.adjust_pointer);
    keep(&mut o.adjust_speed, &dto.adjust_speed);

    // READ-ONLY legacy: the app-level values are applied first, and a layout that still
    // carries its own overrides them so nobody's priority changes on upgrade. Nothing
    // writes them back (see `to_layout_options_dto`), so this ends at the first save.
    keep(&mut o.priority, &dto.priority);
    keep(&mut o.priority_unhooked, &dto.priority_unhooked);
}

/// C# `Apply(PhysicalMonitorModel, ModelDto)`: a model's stored size, borders and name.
pub fn apply_model(layout: &mut Layout, pnp_code: &str, dto: &ModelDto) {
    layout.edit_model(pnp_code, |size, name| {
        let fixed_ratio = size.fixed_aspect_ratio();
        size.set_fixed_aspect_ratio(false);

        let borders = dto.borders.as_ref();
        size.set_top_border(borders.and_then(|b| b.top).unwrap_or(size.borders().top));
        size.set_right_border(
            borders
                .and_then(|b| b.right)
                .unwrap_or(size.borders().right),
        );
        size.set_bottom_border(
            borders
                .and_then(|b| b.bottom)
                .unwrap_or(size.borders().bottom),
        );
        size.set_left_border(borders.and_then(|b| b.left).unwrap_or(size.borders().left));

        // Placeholder sizes (#419) and pre-5.4.1 oriented sizes (#507).
        let (width, height) = layout_migrations::stored_model_size(
            size.width(),
            size.height(),
            dto.width,
            dto.height,
        );
        if let Some(w) = width {
            size.set_width(w);
        }
        if let Some(h) = height {
            size.set_height(h);
        }

        size.set_fixed_aspect_ratio(fixed_ratio);

        if let Some(pnp_name) = dto.pnp_name.as_ref().filter(|n| !n.is_empty()) {
            *name = Some(pnp_name.clone());
        }
    });
}

/// C# `Apply(PhysicalMonitor, MonitorDto)`: a monitor's stored placement, sources,
/// border resistance, exclusion and own borders.
pub fn apply_monitor(layout: &mut Layout, id: &str, dto: &MonitorDto) {
    let Some(sources) = layout.monitor(id).map(|m| m.sources.clone()) else {
        return;
    };
    for source_id in &sources {
        let Some(stored) = dto.sources.as_ref().and_then(|s| s.get(source_id)) else {
            continue;
        };
        let Some(source) = layout.source(source_id).map(|s| s.source.clone()) else {
            continue;
        };

        // Detached sources restore their stored pixel geometry (nothing current to
        // keep); attached ones keep the live geometry, the store is just a backup.
        if !source.attached_to_desktop {
            let px = source.in_pixel;
            layout.set_source_in_pixel(
                source_id,
                Rect::from_location_size(
                    Point::new(
                        stored.pixel_x.unwrap_or(px.x),
                        stored.pixel_y.unwrap_or(px.y),
                    ),
                    Size::new(
                        stored.pixel_width.unwrap_or(px.width),
                        stored.pixel_height.unwrap_or(px.height),
                    ),
                ),
            );
            layout.set_source_orientation(
                source_id,
                stored.orientation.unwrap_or(source.orientation),
            );
        }

        if dto.active_source.as_deref() == Some(source_id.as_str()) {
            layout.set_active_source(id, Some(source_id.clone()));
        }
    }

    if let Some(x) = dto.x_location_in_mm {
        let y = location(layout, id).y;
        layout.set_location(id, Point::new(x, y));
        layout.set_placed(id, true);
    }
    if let Some(y) = dto.y_location_in_mm {
        let x = location(layout, id).x;
        layout.set_location(id, Point::new(x, y));
        layout.set_placed(id, true);
    }

    let ratio = layout
        .monitor(id)
        .map_or(Ratio::uniform(1.0), |m| m.depth_ratio);
    layout.set_depth_ratio(
        id,
        Ratio::new(
            dto.physical_ratio_x.unwrap_or(ratio.x),
            dto.physical_ratio_y.unwrap_or(ratio.y),
        ),
    );

    // The edge length is needed to convert a stored whole-edge resistance into the
    // section that now expresses it. C# reads it off the depth projection, which a
    // monitor without an active source does not have (C# would throw there): no
    // length, and the legacy value is dropped.
    let (across_mm, down_mm) = layout
        .monitor(id)
        .and_then(|m| layout.depth_projection(m))
        .map_or((0.0, 0.0), |dp| (dp.width, dp.height));

    let resistance = dto.border_resistance.as_ref();
    layout.edit_border_resistance(id, |br| {
        apply_side(
            &mut br.left,
            resistance.and_then(|r| r.left.as_ref()),
            down_mm,
        );
        apply_side(
            &mut br.top,
            resistance.and_then(|r| r.top.as_ref()),
            across_mm,
        );
        apply_side(
            &mut br.right,
            resistance.and_then(|r| r.right.as_ref()),
            down_mm,
        );
        apply_side(
            &mut br.bottom,
            resistance.and_then(|r| r.bottom.as_ref()),
            across_mm,
        );
    });

    if let Some(excluded) = dto.excluded_from_layout {
        layout.set_excluded(id, excluded);
    }

    // Per-monitor bezel borders load whatever the current mode is, so switching to
    // PerMonitor is live (no restart required). Stored values only exist once the user
    // edited them in PerMonitor mode: until then the borders keep mirroring the live
    // model values, so the FIRST switch starts from the monitor's current PerModel
    // borders.
    if let Some(b) = &dto.borders {
        let current = layout
            .monitor(id)
            .map_or(Thickness::default(), Monitor::borders);
        layout.set_monitor_borders(
            id,
            Thickness::new(
                b.left.unwrap_or(current.left),
                b.top.unwrap_or(current.top),
                b.right.unwrap_or(current.right),
                b.bottom.unwrap_or(current.bottom),
            ),
        );
        layout.set_borders_customized(id, true);
    }
}

fn location(layout: &Layout, id: &str) -> Point {
    layout
        .monitor(id)
        .map_or(Point::new(0.0, 0.0), Monitor::location)
}

/// C# `Apply(BorderSide, BorderSideDto?, double)`: an edge's sections, a stored
/// whole-edge resistance migrated into one (see [`layout_migrations::sections`]).
pub fn apply_side(side: &mut BorderSide, dto: Option<&BorderSideDto>, edge_length_mm: f64) {
    let Some(dto) = dto else { return };

    side.sections = layout_migrations::sections(dto, edge_length_mm)
        .iter()
        .map(|s| {
            BorderSection::new(
                s.from.unwrap_or(0.0),
                s.to.unwrap_or(0.0),
                s.r#move.unwrap_or(0.0),
                s.move_block.unwrap_or(false),
                s.drag.unwrap_or(0.0),
                s.drag_block.unwrap_or(false),
            )
        })
        .collect();
}

//==================//
// model -> DTO     //
//==================//

/// C# `ToGlobalOptionsDto`: the app-level options. `excluded_defaults_version` is not
/// part of the options model: it is read at load time and round-tripped through every
/// write so the one-time exclusion-list top-up stays one-time.
pub fn to_global_options_dto(
    o: &LayoutOptions,
    excluded_defaults_version: Option<i32>,
) -> GlobalOptionsDto {
    GlobalOptionsDto {
        priority: Some(o.priority.clone()),
        priority_unhooked: Some(o.priority_unhooked.clone()),
        home_cinema: Some(o.home_cinema),
        pinned: Some(o.pinned),
        auto_update: Some(o.auto_update),
        start_minimized: Some(o.start_minimized),
        start_elevated: Some(o.start_elevated),
        debug_tools: Some(o.debug_tools),
        experimental_features: Some(o.experimental_features),
        vcp_control: Some(o.vcp_control),
        show_monitor_action_warning: Some(o.show_monitor_action_warning),
        border_values: Some(o.border_values.clone()),
        rescue_shortcut: Some(o.rescue_shortcut.clone()),
        hide_tray_icon: Some(o.hide_tray_icon),
        excluded_defaults_version,
        ..Default::default()
    }
}

/// C# `ToLayoutDto`: the layout document, its monitors in the layout's order.
pub fn to_layout_dto(layout: &Layout) -> LayoutDto {
    LayoutDto {
        options: Some(to_layout_options_dto(&layout.options)),
        monitors: layout
            .monitors()
            .iter()
            .map(|m| (m.id.clone(), to_monitor_dto(layout, m)))
            .collect(),
        ..Default::default()
    }
}

/// C# `ToDto(ILayoutOptions)`: the per-layout options.
///
/// Priority/PriorityUnhooked are deliberately absent: they are app-level options that
/// once lived here too, and both locations load into the SAME options property. Writing
/// the property back to both made a layout's leftover value the app-level one at the
/// next save, and gave a copy of it to every other layout after that. They are written
/// to the app level only; the layout copy is read as a fallback and then dropped.
pub fn to_layout_options_dto(o: &LayoutOptions) -> LayoutOptionsDto {
    LayoutOptionsDto {
        allow_overlaps: Some(o.allow_overlaps),
        allow_discontinuity: Some(o.allow_discontinuity),
        algorithm: Some(o.algorithm.clone()),
        minimal_edge_overlap: Some(o.minimal_edge_overlap),
        max_travel_distance: Some(o.max_travel_distance),
        freelook_check_interval: Some(o.freelook_check_interval),
        freelook_enabled: Some(o.freelook_enabled),
        loop_x: Some(o.loop_x),
        loop_y: Some(o.loop_y),
        enabled: Some(o.enabled),
        adjust_pointer: Some(o.adjust_pointer),
        adjust_speed: Some(o.adjust_speed),
        ..Default::default()
    }
}

/// C# `ToDto(BorderSide)`: sections only. The per-edge resistance is gone, and writing
/// it back would resurrect it on the next load through the migration path. The side is
/// always emitted, even empty, so deleting the last section clears what the store
/// still holds.
pub fn to_side_dto(side: &BorderSide) -> BorderSideDto {
    BorderSideDto {
        sections: (!side.sections.is_empty()).then(|| {
            side.sections
                .iter()
                .map(|s| BorderSectionDto {
                    from: Some(s.from()),
                    to: Some(s.to()),
                    r#move: Some(s.move_resistance()),
                    move_block: Some(s.move_block()),
                    drag: Some(s.drag()),
                    drag_block: Some(s.drag_block()),
                    ..Default::default()
                })
                .collect()
        }),
        ..Default::default()
    }
}

/// C# `ToDto(PhysicalMonitor)`: a monitor's placement, resistance, own borders and
/// attached sources.
pub fn to_monitor_dto(layout: &Layout, monitor: &Monitor) -> MonitorDto {
    let location = monitor.location();
    let br = &monitor.border_resistance;
    let borders = monitor.borders();
    MonitorDto {
        x_location_in_mm: Some(location.x),
        y_location_in_mm: Some(location.y),
        physical_ratio_x: Some(monitor.depth_ratio.x),
        physical_ratio_y: Some(monitor.depth_ratio.y),
        border_resistance: Some(BorderResistanceDto {
            left: Some(to_side_dto(&br.left)),
            top: Some(to_side_dto(&br.top)),
            right: Some(to_side_dto(&br.right)),
            bottom: Some(to_side_dto(&br.bottom)),
            ..Default::default()
        }),
        // Stored whatever the current mode is (they must survive a save made in
        // PerModel mode), but only once the monitor owns them: uncustomized monitors
        // keep mirroring the model and store nothing.
        borders: monitor.borders_customized().then(|| BordersDto {
            left: Some(borders.left),
            top: Some(borders.top),
            right: Some(borders.right),
            bottom: Some(borders.bottom),
            ..Default::default()
        }),
        active_source: monitor.active_source.clone(),
        serial_number: monitor.serial_number.clone(),
        excluded_from_layout: Some(monitor.excluded_from_layout),
        sources: Some(
            monitor
                .sources
                .iter()
                .filter_map(|id| layout.source(id))
                .filter(|s| s.source.attached_to_desktop)
                .map(|s| {
                    let d = &s.source;
                    let dto = SourceDto {
                        pixel_x: Some(d.in_pixel.x),
                        pixel_y: Some(d.in_pixel.y),
                        pixel_width: Some(d.in_pixel.width),
                        pixel_height: Some(d.in_pixel.height),
                        orientation: Some(d.orientation),
                        display_name: d.display_name.clone(),
                        primary: Some(d.primary),
                        ..Default::default()
                    };
                    (d.id.clone(), dto)
                })
                .collect(),
        ),
        ..Default::default()
    }
}

/// C# `ToDto(PhysicalMonitorModel)`: a model's size, borders and name.
pub fn to_model_dto(model: &MonitorModel) -> ModelDto {
    let size = &model.physical_size;
    let borders = size.borders();
    ModelDto {
        width: Some(size.width()),
        height: Some(size.height()),
        borders: Some(BordersDto {
            left: Some(borders.left),
            top: Some(borders.top),
            right: Some(borders.right),
            bottom: Some(borders.bottom),
            ..Default::default()
        }),
        pnp_name: model.pnp_device_name.clone(),
        ..Default::default()
    }
}

/// The models of a layout's monitors, once each, in the order the monitors first use
/// them (C#: `Select(m => m.Model).DistinctBy(m => m.PnpCode).ToDictionary(...)`).
pub fn to_model_dtos(layout: &Layout) -> IndexMap<String, ModelDto> {
    let mut models = IndexMap::new();
    for monitor in layout.monitors() {
        if !models.contains_key(&monitor.model) {
            if let Some(model) = layout.model(&monitor.model) {
                models.insert(monitor.model.clone(), to_model_dto(model));
            }
        }
    }
    models
}
